Avoid `RangeError` without explicit `limit`

kamipo opened the pull request:
Raising `RangeError` was introduced in order to ease to handle out of
range numbers for `find` and `find_by` in postgresql adapter with
prepared statements since #17459. At that time, `Type::Integer` assumed
implicit default byte size is 4 bytes due to `extract_limit` in
postgresql adapter did not extract an integer byte size correctly.
That assumption is harmful for other adapters and caused #22594.
In #26386, it is fixed to register integer types limit correctly.
The handling whether out of range or not is only needed with explicit
`limit`, and `Type::BigInteger` is no longer needed.

Fixes #22594.

matthewd commented:
@kamipo this seems to be an architectural/philosophical change, not a direct bug fix.

There's nothing wrong with that, but it warrants some more description and explanation. I don't think we should do this, for mostly the same reasons I don't think we should do #25230... but I'm open to being convinced.

kamipo committed:
Avoid `RangeError` without explicit `limit`

Raising `RangeError` was introduced in order to ease to handle out of
range numbers for `find` and `find_by` in postgresql adapter with
prepared statements since #17459. At that time, `Type::Integer` assumed
implicit default byte size is 4 bytes due to `extract_limit` in
postgresql adapter did not extract an integer byte size correctly.
That assumption is harmful for other adapters and caused #22594.
In #26386, it is fixed to register integer types limit correctly.
The handling whether out of range or not is only needed with explicit
`limit`, and `Type::BigInteger` is no longer needed.

kamipo commented:
@matthewd Thank you for your comment. I added more explanatory comments for this PR. WDYT?

kamipo committed:
Implicit `DEFAULT_LIMIT` should be 8 bytes (64-bit signed integer)

SQLite integer is stored in 1, 2, 3, 4, 6, or 8 bytes depending on the
magnitude of the value. Assuming as 4 bytes causes that valid in range
value cannot be stored and existing value cannot be found.

https://www.sqlite.org/datatype3.html

Fixes #22594.

namusyaka commented:
Why has not this been merged?

namusyaka commented:
Amazing. I would love to use this change.

Farranco commented:
I am having issues with this in [Padrino and MySQL adapter](https://stackoverflow.com/questions/43431244/type-casting-a-column-to-bigint-in-padrino-active-record). Would love a fix.

kamipo commented:
Closing this in favor of #28379 and #30000.
